content/blog/2019-11-13-llvm-profiling/index.md

```diff
@@ -0,0 +1,336 @@
+        switch (it->getOpcode()) {
+            case Instruction::Mul:// multiplication
+                mul_instr++;
+                continue;
+            case Instruction::Br:// branch
+                br_instr++;
+                continue;
+            case Instruction::Store:// store
+            case Instruction::Load:// load
+                mem_instr++;
+                continue;
+            default:
+                break;
+        }
+    }
+```
+
+### Function calls
+
+We profile the number of times each function was called in the program. This is done by first initializing global variables corresponding to all the functions in the program to zero. This is done statically by iterating over a function list in the module given by `getFunctionList()`. This returns all the functions called in the program including the ones that were included from C libraries like `printf` or `scanf`. 
+
+```c
+auto &functionList = M.getFunctionList(); // gets the list of functions
+for (auto &function : functionList) { //iterates over the list
+    Value *atomic_counter = new GlobalVariable(M, I64Ty, false, GlobalValue::CommonLinkage, ConstantInt::get(I64Ty, 0),  function.getName()+".glob"); // create a global variable, name it based on the function name
+} 
+
+```
+
+Next we want to insert atomic counters at the start of each function definition. This ensures that irrespective of multiple return points in the function we can always increment the counter for it at the beginning. We start with the entry basic block in the function which is given by the iterator `F.begin()`.  To insert it at the top of the basic block we use `getFirstNonPHI()` which returns the first instruction, which is not a PHI node. We insert an atomic add instruction similar to other profiling instructions. 
+
+
+
+
+
+
+## Hardest Parts
+
+1. For people who are new to LLVM, instructions are hard to find and follow. Searching on google can help if you know what you're looking for. It's difficult to get helpful information unless our search is confined to existing functions. Even though LLVM documentation is pretty exhaustive, it has too many functions to go through and the lack of examples can be off putting. Tutorials and existing backbone codes on Github can be really handy in these scenarios, which we took advantage of. It not only helped us implement specific functions like `printf` but also establish a structure to our IR pass. 
+2. String manipulations: I am not sure if this is the right term to use, but LLVM seems to have 2 string types - Twine and StringRef. `getName` on a function returns a StringRef. In order to make a custom name I perform `F.getName()+"name"` which returns a Twine. But the function `getGlobalVariable` only accepts StringRef. Twine has a function `str` which can be used to convert it into string. Even though this is a straight forward solution, it ended up taking time to figure out the problem and looking into the documentation of these classes.
```

Comment:
Maybe this is already clear, but there's an important difference between `Twine` (which is an actual string object containing actual characters, like a C++ `std::string`) and `StringRef` (which is just a pointer to characters stored elsewhere, like a C `char*`). When you delete a `Twine`, any other pointers to the same data become invalid. When you delete a `StringRef`, nothing really happens and no character memory is freed.